### PR TITLE
Specify merkle value

### PR DESCRIPTION
**Pull request by bkchr**
Closes: https://github.com/paritytech/json-rpc-interface-spec/issues/88

### CODE REVIEW

File: src/api.md [hunk 1]
```diff
@@ -8,3 +8,10 @@ Any missing parameter, or parameter with an invalid format, should result in a J
 
 - "hexadecimal-encoded" designates a binary value encoded as hexadecimal. The value must either be empty, or start with `"0x"` and contain an even number of characters.
 - "SCALE-encoded" designates a value encoded using [the SCALE codec](https://docs.substrate.io/v3/advanced/scale-codec/).
+- ["Merkle value"](https://spec.polkadot.network/chap-state#defn-merkle-value) designates a SCALE encoded value of the following type:
+  ```
+  enum MerkleValue {
+      Node(Vec<u8>),
+      Hash(Hash)
+  }
+  ```
```
[the remark this answers is not in the document]
Comment (bkchr) [reply src/api.md:17]: Thank you for explaining me how the type would be SCALE encoded.

The point being that if we want to make this non opaque, it means that we would need to have a fixed format to decode this. As I said in the issue, the value is either a 32 bytes long or the hash is 32 bytes long. You can not distinguish them. You need to have some form of encoding to distinguish them.

[the remark this answers is not in the document]
Comment (tomaka) [reply src/api.md:17]: What I said in #88 was wrong indeed, but there's no need for the JSON-RPC client to know whether this is a node value or a hash. From the point of view of the JSON-RPC client it's opaque.
The only thing that matters is that all JSON-RPC servers generate the same value in a deterministic way, and for that there's no need for that enum.

[the remark this answers is not in the document]
Comment (bkchr) [reply src/api.md]: I assume you will need https://github.com/polkadot-fellows/RFCs/pull/9 to implement this in smoldot?

[the remark this answers is not in the document]
Comment (tomaka) [reply src/api.md]: No, it already works right now.
If the JSON-RPC client asks for the "closest descendant merkle value" of `foo`, I ask for a proof of `parent(foo)`.

File: src/api.md [hunk 2]
```diff
@@ -8,3 +8,4 @@ Any missing parameter, or parameter with an invalid format, should result in a J
 
 - "hexadecimal-encoded" designates a binary value encoded as hexadecimal. The value must either be empty, or start with `"0x"` and contain an even number of characters.
 - "SCALE-encoded" designates a value encoded using [the SCALE codec](https://docs.substrate.io/v3/advanced/scale-codec/).
+- "Merkle value" as decribed in the [Polkadot spec](https://spec.polkadot.network/chap-state#defn-merkle-value).
```
Comment (tomaka) [inline src/api.md:11]: I don't think that's really the right place, as the Merkle value is used by a single function in the entire API. I'd just put that sentence in `chainHead_unstable_follow.md`.

File: src/api/chainHead_unstable_follow.md [hunk 3]
```diff
@@ -200,7 +200,7 @@ The `value` field is set if this item corresponds to one of the requested items
 
 The `hash` field is set if this item corresponds to one of the requested items whose `type` was `"hash"` or `"descendantsHashes"`. The `hash` field is a string containing the hexadecimal-encoded hash of the storage entry.
 
-The `closestDescendantMerkleValue` field is set if this item corresponds to one of the requested items whose `type` was `"closestDescendantMerkleValue"`. The trie node whose Merkle value is indicated in `closestDescendantMerkleValue` is not indicated, as determining the key of this node might incur an overhead for the JSON-RPC server.
+The `closestDescendantMerkleValue` field is set if this item corresponds to one of the requested items whose `type` was `"closestDescendantMerkleValue"`. The trie node whose Merkle value is indicated in `closestDescendantMerkleValue` is not indicated, as determining the key of this node might incur an overhead for the JSON-RPC server. The value returned by the server must be either the node value or the hash of the node as defined by the Merkle value.
```
Comment (tomaka) [inline src/api/chainHead_unstable_follow.md:203]:
```suggestion
The `closestDescendantMerkleValue` field is set if this item corresponds to one of the requested items whose `type` was `"closestDescendantMerkleValue"`. The trie node whose Merkle value is indicated in `closestDescendantMerkleValue` is not indicated, as determining the key of this node might incur an overhead for the JSON-RPC server. The Merkle value is equal to either the node value or the hash of the node value, as defined in the [Polkadot specification](https://spec.polkadot.network/chap-state#defn-merkle-value).
```

File: src/api.md (hunk 2 above)
Comment (bkchr) [reply src/api.md:11]: You also mention it in `chainHead_unstable_storage`. IMO would be nice to just use the "global" glossary.

**Comment by bkchr** [comment]
@tomaka are you happy with the current state?

**Review by josepot — APPROVED** [review]

File: src/api.md (hunk 2 above)
Comment (tomaka) [inline src/api.md:11]:
```suggestion
- "Merkle value" is described in the [Polkadot specification](https://spec.polkadot.network/chap-state#defn-merkle-value).
```

**Review by tomaka — APPROVED** [review]

**Pull request merged**